src/core/createCookie.js

```diff
@@ -23,58 +23,85 @@ const safeJSONParse = (object, cookieName) => {
   }
 };
 
+/**
+ * The purpose of this proxy is to cache the cookie so we don't have to
+ * read and deserialize it every time a piece of it is accessed.
+ */
+const createCookieProxy = (propertyID, cookieDomain = "") => {
+  const cookieName = `${ALLOY_COOKIE_NAME}_${propertyID}`;
+
+  let deserializedCookie;
+  let cookieHasBeenRead = false;
+
+  return {
+    get() {
+      // We don't read the cookie right when the cookie proxy is created
+      // because we don't know if the user has opted in. If the user
+      // hasn't opted in, we're legally obligated to not read the cookie.
+      // If a component tries to read something off the cookie though,
+      // we assume that the component has received word that the user
+      // has opted-in. The responsibility is on the component.
+      if (!cookieHasBeenRead) {
+        const serializedCookie = cookie.get(cookieName);
+        deserializedCookie =
+          serializedCookie && safeJSONParse(serializedCookie, cookieName);
+        cookieHasBeenRead = true;
+      }
+      return deserializedCookie;
+    },
+    set(updatedCookie) {
+      deserializedCookie = updatedCookie;
+      cookie.set(cookieName, updatedCookie, {
+        expires: ALLOY_COOKIE_TTL,
+        domain: cookieDomain || getTopLevelCookieDomain(window, cookie)
+      });
+    }
+  };
+};
```

Comment:
This is an optimization so that we're not reading+deserializing the cookie every time `cookie.get()` is called, which makes it more amenable for the Identity component to call it every time within its `onBeforeEvent` and `onBeforeDataCollection`.